tests/qos/test_buffer.py

```diff
@@ -0,0 +1,475 @@
+        logging.info('[testcase: headroom override => dynamic headroom]')
+        duthost.shell(set_command)
+        check_profile(duthost, buffer_pg, expected_profile)
+
+        # remove all static profiles
+        logging.info('[restoring configuration]')
+        duthost.shell('config buffer profile remove headroom-override')
+        duthost.shell('config buffer profile remove non-default-dynamic_th')
+        check_lossless_profile_removed(duthost, 'headroom-override')
+        check_lossless_profile_removed(duthost, 'non-default-dynamic_th')
+    finally:
+        if pg_to_test == '3-4':
+            duthost.shell(set_command, module_ignore_errors = True)
+        else:
+            duthost.shell('config interface buffer priority-group lossless remove {} {} '.format(port_to_test, pg_to_test), module_ignore_errors = True)
+        duthost.shell('config interface cable-length {} {}'.format(port_to_test, original_cable_len), module_ignore_errors = True)
+        duthost.shell('config buffer profile remove headroom-override', module_ignore_errors = True)
+        duthost.shell('config buffer profile remove non-default-dynamic_th', module_ignore_errors = True)
+
+
+def test_exceeding_headroom(duthost, conn_graph_facts, port_to_test):
+    '''
+    @summary: 
+    '''
+    max_headroom_size = duthost.shell('redis-cli -n 6 hget "BUFFER_MAX_PARAM_TABLE|{}" max_headroom_size'.format(port_to_test))['stdout']
+    if not max_headroom_size:
+        pytest.skip('No max headroom found on port {}, skip'.format(port_to_test))
+
+    original_cable_len = duthost.shell('redis-cli -n 4 hget "CABLE_LENGTH|AZURE" {}'.format(port_to_test))['stdout']
+    original_speed = duthost.shell('redis-cli -n 4 hget "PORT|{}" speed'.format(port_to_test))['stdout']
+    original_profile = 'pg_lossless_{}_{}_profile'.format(original_speed, original_cable_len)
+
+    try:
+        # set to super long cable length
+        logging.info('[config a super long cable length]')
+        duthost.shell('config interface cable-length {} 10000m'.format(port_to_test))
+        time.sleep(20)
+        logging.info('verify the profile isn\'t changed')
+        check_profile(duthost, 'BUFFER_PG_TABLE:{}:3-4'.format(port_to_test), original_profile)
+        duthost.shell('config interface cable-length {} {}'.format(port_to_test, original_cable_len))
+
+        # add additional PG
+        logging.info('[config the cable length on the port]')
+        duthost.shell('config interface cable-length {} 300m'.format(port_to_test))
+        time.sleep(20)
+        logging.info('verify the profile has been changed')
+        expected_profile = 'pg_lossless_{}_{}_profile'.format(original_speed, '300m')
+        check_profile(duthost, 'BUFFER_PG_TABLE:{}:3-4'.format(port_to_test), expected_profile)
+        logging.info('add another PG and make sure the system isn\'t broken')
+        duthost.shell('config interface buffer priority-group lossless add {} {}'.format(port_to_test, '5-7'))
```

Comment:
How will this work when these are already configured as lossy PG's?

Reply:
There will be error in the log and the configuration won’t be applied.
In terms of regression test, it should not hit this situation since only pg1 is configured as lossy.